### PR TITLE
experimental: multi-column notebooks

### PR DESCRIPTION
**Pull request by wasimsandhu**
## 📝 Summary

Enables multi-column functionality for marimo notebooks! 🚀

Screenshot 2024-10-07 at 6 33 11 PM

## 🔍 Description of Changes

- [x] Toggle columns by selecting the "columns" option in Notebook Menu > App Width
- [x] Represent columns in frontend state model by organizing `state.cellIds` with `MultiColumn<CellId>`, an array `CollapsibleTree<CellId>` objects.
- [x] Add/delete columns using the `+` or `x` button in the column header
- [x] Insert column breakpoints on cells
- [x] Serialize columns into notebook file using new `column` property in cell config
- [x] Update cell movement actions and drag-and-drop across columns
- [x] Update and add unit tests for columns

## 📋 Checklist

- [x] I have read the [contributor guidelines](https://github.com/marimo-team/marimo/blob/main/CONTRIBUTING.md).
- [x] For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://discord.gg/JE7nhX6mD8), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [x] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

## 📜 Reviewers

@akshayka OR @mscolnick

### CODE REVIEW

**Comment by wasimsandhu** [comment]
Might just be me, but I think we could use some more contrast between the column header and app background

Screenshot 2024-10-07 at 7 43 19 PM

**Comment by mscolnick** [comment]
@wasimsandhu maybe we can increase the slate scale for both the body and header by 1 of the column wrapper? Does that look better?

**Comment by wasimsandhu** [comment]
@mscolnick Just tried it, but I think I've been looking at it too long. Might be good to have some fresh eyes on it

File: frontend/src/core/hotkeys/hotkeys.ts [hunk 1]
```diff
@@ -81,6 +81,11 @@ const DEFAULT_HOT_KEY = {
     group: "Creation and Ordering",
     key: "Mod-Shift-2",
   },
+  "cell.addColumnBreakpoint": {
+    name: "Add column breakpoint",
+    group: "Creation and Ordering",
+    key: "Mod-Shift-3",
```
Comment (wasimsandhu) [inline frontend/src/core/hotkeys/hotkeys.ts:87]: What should the hotkey for this be? Cmd+Shift+3 is already taken for screenshots on MacOS

File: frontend/src/utils/arrays.ts [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
 }
 
 export function arrayInsert<T>(array: T[], index: number, value: T): T[] {
+  index = clamp(index, 0, array.length);
   return arrayInsertMany(array, index, [value]);
 }
 
+export function arrayMove<T>(array: T[], from: number, to: number): T[] {
+  const value = array[from];
+  return arrayInsertMany(arrayDelete(array, from), to, [value]);
```
Comment (wasimsandhu) [inline frontend/src/utils/arrays.ts:16]: Nice

**Review by wasimsandhu — COMMENTED** [review]
LGTM

File: frontend/src/core/config/config-schema.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ export const PackageManagerNames = [
 ] as const;
 export type PackageManagerName = (typeof PackageManagerNames)[number];
 
-export const APP_WIDTHS = ["compact", "medium", "full"] as const;
+export const APP_WIDTHS = ["compact", "medium", "full", "columns"] as const;
```
[the remark this answers is not in the document]
Comment (mscolnick) [reply frontend/src/core/config/config-schema.ts:17]: you might be running an old version of marimo? maybe in a sandbox?

File: frontend/src/core/hotkeys/hotkeys.ts (hunk 1 above)
Comment (mscolnick) [reply frontend/src/core/hotkeys/hotkeys.ts:87]: `Mod+Shift+/`?

**Review by akshayka — COMMENTED** [review]
High-level thoughts:

I can definitely see this being useful for exploratory work, and for keeping long notebooks organized! I would personally enjoy writing my pure functions in one column and calling them in another, as I've said before.

I could see this introducing friction between collaborators who have monitors of differing widths; but I can think of ways of improving that experience in the future, such as having canvas like zoom.

We originally talked about releasing under `experimental` and advertising it on Discord to get more feedback -- do we still want to do that?

Granular feedback:

1. Run button style regression: When a cell is stale, the run button should be highlighted with its stale background.

This PR (incorrect):

image

For the correct styling see what happens on `0.9.3`.

2. Elapsed time string clips out of the column; coupled with staleness indicator it clips out of the column even more.

3. Can we add tooltips to the column header buttons? I was using a small screen, and I couldn't figure out what the `<` and `>` buttons were doing.

4. I'm all for compactness, but these cells feel a bit narrow even for me.

5. When a new column is created, it should be populated with an empty cell.

6. Is the `min-height` of `400px` needed, or can it be decreased? For a new column it positions the "Python, Markdown, SQL, AI" buttons far from the new cell.

**Comment by mscolnick** [comment]
I've moved this to a feature flag, can be added by adding this to your marimo.toml:
```
[experimental]
multi_column=true
```

**Review by mscolnick — DISMISSED** [review]

**Review by akshayka — DISMISSED** [review]

**Pull request merged**